NewForce-at-Mountwest/bangazon-api-the-thunder-lobsters
# Added Orders Resource

# Description
Added functionality to GET, POST, PUT, and DELETE Orders from the Bangazon API. Added a test suite for integration testing of this functionality. Added string queries to the GET methods for this resource.


## Related Ticket(s)
https://trello.com/c/79sCrlHp/13-ticket-4-allow-developers-to-access-the-order-resource


## Steps to Test
- Pull down these changes on a new branch
- Open a new SQL query for your database
- Copy the following code into the new SQL query and execute it:

`INSERT INTO PaymentType (AcctNumber, Name, CustomerId) VALUES ('1', 'Payment One', (SELECT TOP 1 Id FROM Customer WHERE FirstName = 'Potato'));`

`INSERT INTO [Order] (CustomerId, PaymentTypeId) VALUES ((SELECT TOP 1 Id FROM Customer WHERE FirstName = 'Potato'), (SELECT TOP 1 Id FROM PaymentType WHERE Name = 'Payment One'));`

`INSERT INTO [Order] (CustomerId, PaymentTypeId) VALUES ((SELECT TOP 1 Id FROM Customer WHERE FirstName = 'Potato'), (SELECT TOP 1 Id FROM PaymentType WHERE Name = 'Payment One'));`

`INSERT INTO OrderProduct (OrderId, ProductId) VALUES ((SELECT TOP 1 [Order].Id FROM [Order] JOIN Customer ON [Order].CustomerId = Customer.Id WHERE Customer.FirstName = 'Potato'), (SELECT TOP 1 Id FROM Product WHERE Title = 'Test One'));`

`INSERT INTO [Order] (CustomerId) VALUES ((SELECT TOP 1 Id FROM Customer WHERE FirstName = 'Potato'));`

- If you don't have an `appsettings.json` file, create one in the BangazonAPI project and paste the following code:
`{
 "Logging": {
    "LogLevel": {
      "Default": "Warning"
    }
  },
  "AllowedHosts": "*",
  "ConnectionStrings": {
    "DefaultConnection": "Server=localhost\\SQLEXPRESS01;Database=BangazonAPI;Trusted_Connection=True;"
  }
}`

- Edit `DatabaseFixture.cs` (in TestBangazonAPI) to reference the SQL server running on your machine, rather than SQLEXPRESS01.
- Press CTRL + F5 to start the API
- Change the URL of your browser to end in api/orders
- Copy and paste this URL into Postman
- Using Postman, ensure that you can GET a list of all products from this URL. This should include at least one Order resource without a PaymentTypeId.
- Ensure that you can add an ID to the end of the URL to GET a single product resource
- Using the default GET method, include `_completed=true`, `_completed=false`, `_include=products`, and `_include=customer` string queries. 
- At least one result should be completed and one should not. One order should have a product, and all orders should have customers.
- Copy some JSON from a GET result and use it to POST a new product
- Use PUT to edit the product you just POSTed
- Use GET to ensure this product was edited
- DELETE the product you created
- Use GET to make sure the product is now removed
- In Visual Studio, navigate to the Test Explorer
- Check the models in `DatabaseFixture.cs` for TestOrder and TestDeleteOrder against your database. Make sure the CustomerId and PaymentTypeId reference the proper entries (PaymentTypeId specifically should point to a PaymentType with a name of "INTEGRATION TEST"). Also check the editedOrder in `OrderTest.cs` for the same ids.
- Run all tests. Ensure that all 5 OrderTest tests succeed.


# Checklist:
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new errors
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
